### PR TITLE
fix(sync): prevent synchronizer loop when very close to tip

### LINKED ISSUE #3375: Fix slowness when syncing near the tip
## Motivation

It seems that sync becomes slow when it's near the tip. Investigate to find the cause and fix it.

There is some discussion about it in the #slow-sync thread in Discord.

### Diagnostics

This is unexpected - the syncer hangs for a few minutes before this log, and there are two chains a long way from the tip:
> Feb 25 03:40:26.085  INFO ***net="Main"***:sync: zebrad::components::sync: extending tips tips.len=2 in_flight=673 lookahead_limit=2000 state_tip=Some(Height(1577565))

https://github.com/ZcashFoundation/zebra/runs/5325705831?check_suite_focus=true#step:7:10962

## Related Work

Possible Causes:
- #862
- #3603

Related Tickets:
- #3438 
- #3042

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Zebra currently has some performance issues when block synchronization is close to the chain tip. During the investigation of the issue, I ran into two causes:

1. The synchronizer cancels all active block downloads when it exhausts its prospective tips set. This results in the final few block downloads being constantly started and cancelled.
2. Timeouts when waiting for inputs from the state service during the validation of transactions.

This PR fixes the first cause, and the second cause will very likely be fixed by the refactor to split the state service to increase throughput of read only requests (#3866).

## Solution

Only cancel active block download tasks in case there's an error, preventing them from being cancelled when the prospective tips set is exhausted.

Closes #3375

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

I'd like to have a regression test for this, but I'll try to open a separate optional PR for it.

### CODE REVIEW

File: zebrad/src/components/sync.rs [hunk 1]
```diff
@@ -777,6 +774,28 @@ where
...
     }
 
+    /// Handles a response for a requested block.
+    ///
+    /// Returns `Ok` if the block was successfully verified and commited to the state, or if an
+    /// expected error occurred, so that the synchronization can continue normally.
+    ///
+    /// Returns `Err` if an unexpected error occurred, to force the synchronizer to restart.
+    async fn handle_block_response(
+        &mut self,
```
Comment (teor2345) [inline zebrad/src/components/sync.rs:784]: Nitpick: `self` is unused and can be deleted, which turns this into an associated function.

**Comment by dconnolly** [comment]
This fixes the issue for me 👍

**Review by teor2345 — APPROVED** [review]
Nice catch!

Thanks for splitting this PR up into commits, it was very helpful during the review.

**Comment by conradoplg** [comment]
Is there something preventing this from moving out of draft?

**Comment by dconnolly** [comment]
@jvff I've been running this since Friday and it's awesome, shall we promote this out of draft?

File: zebrad/src/components/sync.rs (hunk 1 above)
Comment (jvff) [reply zebrad/src/components/sync.rs:784]: I don' think this is necessary, and I like that the `&mut self` semantics here forces callers to immediately await for the result (which helps preserve the original semantics).

We can change it later if there's ever a need for it.

**Pull request merged**